### PR TITLE
enable structuredAttrs for vimPlugins

### PR DESCRIPTION
**Pull request by teto**
- **buildVimPlugin: deprecate addRtp**
- **buildVimPlugin: enable structuredAttrs**

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by teto** [comment]
Should probably target staging 
```
4 packages failed to build:
vimPlugins.hardhat-nvim vimPlugins.idris2-nvim vimPlugins.nvim-dap-ui vimPlugins.sg-nvim

1526 packages built:
```

**Comment by PerchunPak** [comment]
Build logs for failed builds:
https://gist.github.com/PerchunPak/e30576af1044a7639f0d38de0b18735e

**Review by PerchunPak — CHANGES_REQUESTED** [review]
Honestly, I don't know the reason behind those failures, but this should not be merged until this is resolved. We test only a small portion of plugins, which is why I suspect there are even more hidden breakages.

**Comment by teto** [comment]
> Honestly, I don't know the reason behind those failures, but this should not be merged until this is resolved. We test only a small portion of plugins, which is why I suspect there are even more hidden breakages.

These are clearly not linked to this PR. sg.nvim was broken due to the rust bump. The hard-hat one probably by a PR merge without a nixpkgs-review.

File: pkgs/applications/editors/vim/plugins/build-vim-plugin.nix [hunk 1]
```diff
@@ -4,12 +4,8 @@
...
 }:
 
-rec {
-  addRtp = drv:
-    drv // {
-      rtp = lib.warn "`rtp` attribute is deprecated, use `outPath` instead." drv.outPath;
-      overrideAttrs = f: addRtp (drv.overrideAttrs f);
-    };
+{
+  addRtp = drv: lib.warn "`addRtp` is deprecated, does nothing." drv;
```
[the remark this answers is not in the document]
Comment (teto) [reply pkgs/applications/editors/vim/plugins/build-vim-plugin.nix:8]: no one is using addRtp and it breaks overrideAttrs in some ways. It has been deprecated for a few releases now, it's me being nice for the last release and if no one complains about it after next release, remove it completely.

**Comment by PerchunPak** [comment]
> The hard-hat one probably by a PR merge without a nixpkgs-review.

All of them (except for sg.nvim), get built without any issues on master, and only after cherry-picking this PR, there are failures:

```sh
~/dev/nixpkgs @7f871ed07194 ❯ nix build .#vimPlugins.hardhat-nvim .#vimPlugins.idris2-nvim .#vimPlugins.nvim-dap-ui --keep-going --rebuild
fetching git input 'git+file:///home/perchun/dev/nixpkgs'
fetching git input 'git+file:///home/perchun/dev/nixpkgs'
fetching git input 'git+file:///home/perchun/dev/nixpkgs'
checking outputs of '/nix/store/i8zjszpaba1ifbfab7vyygh16sxgyvz9-vimplugin-hardhat.nvim-2024-07-28.drv'...
checking outputs of '/nix/store/kfwava78f9638azfbwgcb58d917asxx4-vimplugin-idris2-nvim-2023-09-05.drv'...
checking outputs of '/nix/store/07jjsfarj6yj7mw9l7wd56qh8dkabnik-vimplugin-nvim-dap-ui-2024-07-13.drv'...
```

**Review by salko-ua — APPROVED** [review]

**Comment by teto** [comment]
> Honestly, I don't know the reason behind those failures, but this should not be merged until this is resolved. We test only a small portion of plugins, which is why I suspect there are even more hidden breakages.

due to structuredAttrs `dependencies` is now passed as an array instead of string, which breaks the neovim test hook bash substitution, ie., neovim sees only one dependency instead of all the dependencies. Should be easy to fix.

**Comment by teto** [comment]
@PerchunPak it fixes build  for hardhat and the one problematic plugins you mentioned (I just built htose, haven't run a nixpkgs-review on it). I think t's ready

**Review by GaetanLepage — APPROVED** [review]
LGTM

**Comment by PerchunPak** [comment]
A bit late, but sacrificed a few GBs of my internet for nixpkgs-review

https://gist.github.com/PerchunPak/55a46813c8e47565e4cf3bbf8e2da49c

**Pull request merged**